crates/bevy_ecs/src/schedule/executor/async.rs

```diff
@@ -0,0 +1,432 @@
+            thread::current(),
+            Arc::new(AtomicI64::new(waker_list_len as i64 - 1)),
+        ));
+        if let None = GLOBAL_WORLD_ACCESS.set(world, || {
+            for waker in waker_list {
+                waker.wake();
+            }
+            thread::park();
+        }) {
+            return None;
+        }
+        world.try_resource_scope(|world, mut appliers: Mut<SystemParamAppliers>| {
+            appliers.run(world);
+        });
+        Some(())
+    }
+}
+
+/// The PhantomData here is just there cause it's a cute way of showing that we have a mutex around our unsafe worldcell and that's what the mutex is 'locking'
+pub(crate) struct WorldAccessRegistry(
+    OnceLock<
+        RwLock<
+            HashMap<
+                WorldId,
+                RwLock<
+                    Option<(
+                        UnsafeWorldCell<'static>,
+                        Mutex<PhantomData<UnsafeWorldCell<'static>>>,
+                    )>,
+                >,
+            >,
+        >,
+    >,
+);
+
+impl WorldAccessRegistry {
+    pub(crate) fn set(&self, world: &mut World, func: impl FnOnce()) -> Option<()> {
+        let this = self.0.get_or_init(|| RwLock::new(HashMap::new()));
+        let world_id = world.id();
+        if !this.read().unwrap().contains_key(&world_id) {
+            // VERY rare only happens the first time we try to do anything async in a new World
+            let _ = this.write().unwrap().insert(world_id, RwLock::new(None));
+        }
+
+        struct ClearOnDropGuard<'a> {
+            slot: &'a RwLock<
+                Option<(
+                    UnsafeWorldCell<'static>,
+                    Mutex<PhantomData<UnsafeWorldCell<'static>>>,
+                )>,
+            >,
+        }
+        impl<'a> Drop for ClearOnDropGuard<'a> {
+            fn drop(&mut self) {
+                // clear it on the way out, even on panic
+                self.slot.write().unwrap().take();
+            }
+        }
+        unsafe {
+            let binding = this.read().unwrap();
+            let world_container = binding.get(&world_id).unwrap();
+            // SAFETY this is required in order to make sure that even in the event of a panic, this can't get accessed
+            let _clear = ClearOnDropGuard {
+                slot: world_container,
+            };
+            // SAFETY: This mem transmute is safe only because we drop it after, and our GLOBAL_WORLD_ACCESS is private, and we don't clone it
+            // where we do use it, so the lifetime doesn't get propagated anywhere.
+            // Lifetimes are not used in any actual code optimization, so turning it into a static does not violate any of rust's rules
+            // As *LONG* as we keep it within it's lifetime, which we do here, manually, with our `ClearOnDrop` struct.
+            world_container.write().unwrap().replace((
+                std::mem::transmute(world.as_unsafe_world_cell()),
```

Comment:
It's good practice to directly annotate the transmute with types, since it's such a powerful operation and it should be both clear at a glance what it is doing and enforced by the compiler rather than allowing type inference to mess anything up.